### PR TITLE
Prevent MySQL error has gone away error during tracking

### PR DESCRIPTION
**Pull request by tsteur**
### Description:

This fixed Cloud-1178 and follows up from https://github.com/matomo-org/matomo/pull/20110

After deploying https://github.com/matomo-org/matomo/pull/20110 we noticed we're already getting less MySQL gone away errors during tracking but were still getting 50-100 per day. This change here reduced it down to around 5 per day and is now basically only still happening for inserts.

Inserts we didn't want to retry to prevent the double execution of a query. Thought to rather not insert than double insert.

This change will now also re-establish the DB connection when:
* there is any kind of select query
* there is any kind of update query
* we try to begin a transaction

It reduces the MySQL Server has gone away errors by 95%.

We don't want these changes in the regular PDO adapter as there would be issues during archiving segments etc.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/Tracker/Db/Pdo/Mysql.php [hunk 1]
```diff
@@ -209,6 +218,57 @@ public function fetch($query, $parameters = array())
...
+
+    /**
+     * @internal for tests only
+     * @param Exception $e
+     * @throws Exception
+     */
+    public function reconnect(Exception $e)
+    {
+        $this->disconnect();
+        usleep(100 * 1000); // wait for 100ms
```
Comment (tsteur) [inline core/Tracker/Db/Pdo/Mysql.php:252]: In tracker we're now using 100ms which we've been testing on production and worked well enough for us

File: core/Tracker/Db/Pdo/Mysql.php [hunk 2]
```diff
@@ -339,14 +412,7 @@ private function establishConnection(): void
             $sql = "SET NAMES '".$this->charset."'";
             $this->connection->exec($sql);
         }
-
-        $this->resetPassword();
     }
 
-    private function resetPassword()
```
Comment (tsteur) [inline core/Tracker/Db/Pdo/Mysql.php:346]: the removal of this shouldn't cause any issues as in other db providers we never did that and it wasn't a problem

File: core/Tracker/Db/Pdo/Mysql.php [hunk 3]
```diff
@@ -209,6 +218,57 @@ public function fetch($query, $parameters = array())
...
+        } catch (Exception $e) {
+            $isSelectQuery = stripos(trim($query), 'select ') === 0;
+            $isUpdateQuery = stripos(trim($query), 'update ') === 0;
+
+            if (($isSelectQuery || $isUpdateQuery)
+                && !$this->activeTransaction
+                && $this->isMysqlServerHasGoneAwayError($e)) {
+                // mysql may return a MySQL server has gone away error when trying to execute the query
+                // in that case we want to retry establishing the connection once after a short sleep
+                // we're only retrying SELECT and UPDATE queries to prevent inserting records twice for some reason
```
Comment (sgiehl) [inline core/Tracker/Db/Pdo/Mysql.php]: Just a note: In theory an update statement could also result in incorrect values when executed twice. Like when increasing or decreasing by certain values.

**Review by sgiehl — APPROVED** [review]
Left a note. Otherwise this looks good to me.
@tsteur feel free to merge if you think that update statements won't cause issues.

File: core/Tracker/Db/Pdo/Mysql.php (hunk 3 above)
Comment (tsteur) [reply core/Tracker/Db/Pdo/Mysql.php]: Good point @sgiehl I'll remove it just to be safe.

**Comment by tsteur** [comment]
Updated this @sgiehl

**Comment by tsteur** [comment]
Maybe have another look to see if we can merge?

**Review by sgiehl — APPROVED** [review]

**Pull request merged**